### PR TITLE
[Merged by Bors] - feat(MorphismProperty/Presheaf): add `relative` morphism property

### PR DESCRIPTION
**Pull request by callesonne**
Given a morphism property P and a functor F, we define the morphism property `P.relative` of morphisms in the codomain of `F`. This is defined as being relatively representable with respect to `F`, and that any represented pullback of the morphism satisfies P.

Co-authored-by: Joël Riou <rioujoel@gmail.com>

---

- [x] depends on: #16140
- [x] depends on: #16141

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by joelriou** [comment]
(I have contributed to this PR, but I just would like to say that this looks good to me.)

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean [hunk 1]
```diff
@@ -288,4 +290,87 @@ instance respectsIso : RespectsIso F.relativelyRepresentable :=
...
+lemma relative_of_snd [F.Faithful] [F.Full] [P.RespectsIso] {f : X ⟶ Y}
+    (hf : F.relativelyRepresentable f) (h : ∀ ⦃a : C⦄ (g : F.obj a ⟶ Y), P (hf.snd g)) :
+    P.relative F f :=
+  relative.of_exists hf (fun _ g ↦ ⟨hf.pullback g, hf.fst g, hf.snd g, hf.isPullback g, h g⟩)
+
+/-- If `P : MorphismProperty C` is stable under base change, `F` is fully faithful and preserves
+pullbacks, and `C` has all pullbacks, then for any `f : a ⟶ b` in `C`, `F.map f` satisfies
+`P.relative` if `f` satisfies `P`. -/
+lemma relative_map [F.Faithful] [F.Full] [PreservesLimitsOfShape WalkingCospan F]
+    [HasPullbacks C] (hP : StableUnderBaseChange P) {a b : C} {f : a ⟶ b}
```
[the remark this answers is not in the document]
Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Representable.lean]: Great suggestion, fixed!

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean [hunk 2]
```diff
@@ -288,4 +291,87 @@ instance respectsIso : RespectsIso F.relativelyRepresentable :=
...
+
+/-- Given a morphism property `P` in a category `C`, a morphism `f : F ⟶ G` of presheaves in the
+category `Cᵒᵖ ⥤ Type v` satisfies the morphism property `P.presheaf` iff:
+* The morphism is representable.
+* For any morphism `g : F.obj a ⟶ G`, the property `P` holds for any represented pullback of
+  `f` by `g`.
+
+This is implemented as a special case of the more general notion of `P.relative`, to the case when
+the functor `F` is `yoneda`. -/
+abbrev presheaf := P.relative yoneda
```
Comment (erdOne) [inline Mathlib/CategoryTheory/MorphismProperty/Representable.lean:318]:
```suggestion
abbrev presheaf : MorphismProperty (Dᵒᵖ ⥤ Type v₂) := P.relative yoneda
```
Not tested

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean [hunk 3]
```diff
@@ -288,4 +291,87 @@ instance respectsIso : RespectsIso F.relativelyRepresentable :=
...
+open Functor.relativelyRepresentable
+
+variable {X Y : D} (P : MorphismProperty C)
+
+/-- Given a morphism property `P` in a category `C`, a functor `F : C ⥤ D` and a morphism
+`f : X ⟶ Y` in `D`. Then `f` satisfies the morphism property `P.relative` with respect to `F` iff:
+* The morphism is representable with respect to `F`
+* For any morphism `g : F.obj a ⟶ Y`, the property `P` holds for any represented pullback of
+  `f` by `g`. -/
+def relative : MorphismProperty D :=
```
Comment (erdOne) [inline Mathlib/CategoryTheory/MorphismProperty/Representable.lean:305]: I feel that the name is a bit too generic but I don't have a better name.

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean [hunk 4]
```diff
@@ -288,4 +291,87 @@ instance respectsIso : RespectsIso F.relativelyRepresentable :=
...
+
+/-- Given a morphism property `P` which respects isomorphisms, then to show that a morphism
+`f : X ⟶ Y` satisfies `P.relative` it suffices to show that:
+* The morphism is representable.
+* For any morphism `g : F.obj a ⟶ G`, the property `P` holds for *some* represented pullback
+of `f` by `g`. -/
+lemma relative.of_exists [F.Faithful] [F.Full] [P.RespectsIso] {f : X ⟶ Y}
+    (hf : F.relativelyRepresentable f) (h₀ : ∀ ⦃a : C⦄ (g : F.obj a ⟶ Y),
+    ∃ (b : C) (fst : F.obj b ⟶ X) (snd : b ⟶ a) (_ : IsPullback fst (F.map snd) f g), P snd) :
+    P.relative F f := by
```
Comment (erdOne) [inline Mathlib/CategoryTheory/MorphismProperty/Representable.lean:343]: I think `hf` is not needed here.

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean (hunk 2 above)
Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Representable.lean:318]: Fixed!

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean (hunk 4 above)
Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Representable.lean:343]: Nice spot!

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean (hunk 3 above)
Comment (callesonne) [reply Mathlib/CategoryTheory/MorphismProperty/Representable.lean:305]: I agree that its generic, but I also don't have a better name for it.

**Comment by erdOne** [comment]
LGTM modulo the comment on naming. Maybe merge it if there aren't better names.
maintainer merge

File: Mathlib/CategoryTheory/MorphismProperty/Representable.lean (hunk 3 above)
Comment (kbuzzard) [reply Mathlib/CategoryTheory/MorphismProperty/Representable.lean:305]: It's stuck away in an obscure namespace so I think this is fine.

**Comment by kbuzzard** [comment]
Great to see progress here :-) There are so many functors we need to represent!

bors merge

**Pull request closed without merging**